map/extrapolation_benchmark/extrapolation_benchmark.cpp

```diff
@@ -0,0 +1,326 @@
+
+DEFINE_string(csv_path, "",
+              "Path to csv file with user in following format: mwm id (string), aloha id (string), "
+              "latitude of the first coord (double), longitude of the first coord (double), "
+              "timestamp in seconds (int), latitude of the second coord (double) and so on.");
+
+using namespace extrapolation;
+using namespace location;
+using namespace routing;
+using namespace std;
+
+namespace
+{
+struct GpsPoint
+{
+  GpsPoint(double timestampS, double lat, double lon)
+      : m_timestampS(timestampS)
+      , m_lat(lat)
+      , m_lon(lon)
+  {
+  }
+
+  double m_timestampS;
+  // @TODO(bykoianko) Using LatLog type instead of two double should be considered.
+  double m_lat;
+  double m_lon;
+};
+
+class Expectation
+{
+public:
+  void Add(double value)
+  {
+    m_averageValue = (m_averageValue * m_counter + value) / (m_counter + 1);
+    ++m_counter;
+  }
+
+  double Get() const { return m_averageValue; }
+  size_t GetCounter() const { return m_counter; }
+
+private:
+  double m_averageValue = 0.0;
+  size_t m_counter = 0;
+};
+
+class ExpectationVec
+{
+public:
+  explicit ExpectationVec(size_t size) { m_mes.resize(size); }
+
+  void Add(vector<double> const & values)
+  {
+    CHECK_EQUAL(values.size(), m_mes.size(), ());
+    for (size_t i = 0; i < values.size(); ++i)
+      m_mes[i].Add(values[i]);
+  }
+
+  vector<Expectation> const & Get() const { return m_mes; }
+
+private:
+  vector<Expectation> m_mes;
+};
+
+using Track = vector<GpsPoint>;
+using Tracks = vector<Track>;
+
+bool GetString(stringstream & lineStream, string & result)
+{
+  if (!lineStream.good())
+    return false;
+  getline(lineStream, result, ',');
+  return true;
+}
+
+bool GetDouble(stringstream & lineStream, double & result)
+{
+  string strResult;
+  if (!GetString(lineStream, strResult))
+    return false;
+  return strings::to_double(strResult, result);
+}
+
+bool GetUint64(stringstream & lineStream, uint64_t & result)
+{
+  string strResult;
+  if (!GetString(lineStream, strResult))
+    return false;
+  return strings::to_uint64(strResult, result);
+}
+
+bool GetGpsPoint(stringstream & lineStream, uint64_t & timestampS, double & lat, double & lon)
+{
+  if (!GetUint64(lineStream, timestampS))
+    return false;
+  if (!GetDouble(lineStream, lat))
+    return false;
+
+  return GetDouble(lineStream, lon);
+}
+
+/// \brief Fills |tracks| based on file |pathToCsv| content. File |pathToCsv| should be
+/// a text file with following lines:
+/// <Mwm name (country id)>, <Aloha id>,
+/// <Latitude of the first point>, <Longitude of the first point>, <Timestamp in seconds of the first point>,
+/// <Latitude of the second point> and so on.
+bool Parse(string const & pathToCsv, Tracks & tracks)
+{
+  tracks.clear();
+
+  std::ifstream csvStream(pathToCsv);
+  if (!csvStream.is_open())
+    return false;
+
+  while (!csvStream.eof())
+  {
+    string line;
+    getline(csvStream, line);
```

Comment:
Why not just:
```c++
string line;
while (getline(csvStream, line)) {
  ...
}
```
?

Reply:
As far as I know getline() returns istream which have operator bool. This operator bool returns true if none of failbit or badbit is set. But I need eofbit.

On the other hand the way you wrote about works as well. Do you know why? :)

Reply:
Why do you need to check eofbit explicitly? Eofbit is set by the getline that was failed to extract any characters, but note that getline also sets failbit in this case. Take a look at http://en.cppreference.com/w/cpp/io/ios_base/iostate.

Reply:
Right. I see. Let's put `while (getline(csvStream, line))`.